### PR TITLE
Track model logp (in HMC and NUTS)

### LINKED ISSUE #2971: Add diagnostic for model logp
In PyMC3 we did not recorde the model logp to the trace. Today I came across this explanation from Betancourt, which I think it is quite excellent and we should consider adding it to the trace or output the diagnoistics: http://discourse.mc-stan.org/t/convergence-failure-maybe-in-brms/4148/7. Copied below:

> lp__ tends to be extremely sensitive to the autocorrelation of the Markov chain and hence provides a reasonable bound on how well any function of the variables will converge. In other words, ensuring that the diagnostics for lp__ are good gives you the strongest evidence that your fit is okay but if you focus only on a few variables and carefully check the diagnostics for those variables then you may be able to ignore lp__ for that very specific context.

### PR DESCRIPTION
**Pull request by eigenfoo**
Close #2971.

WIP. Following up from #3121. I'm opening up this PR so I can get Travis builds/tests/feedback from maintainers 😄 

@ColCarroll I've done a first pass at tracking the logp. The sampler stats stack actually goes deeper than you outlined before, since the model logp is actually kept in `integration.py` for HMC. What I've done is modify the `State` so it keeps track of the model logp, and percolate that up as a sampler stat to HMC and NUTS.

@junpenglao currently this PR falls a bit short of what you were expecting; I've only made the model logp a sampling stat for HMC and NUTS. If I'm not mistaken, no other sampler computes the logp (e.g. Metropolis doesn't), so there would be overhead to require these samplers to track the logp. However, if this is a feature the dev team wants, I'd be happy to do that too - probably in a separate PR, though.

I'm also unsure about the `compoundstep` changes... `states` is a list, and cannot be indexed by a string. I must have misunderstood @junpenglao's comments [here](https://github.com/pymc-devs/pymc3/pull/3121#issuecomment-408788197): I'll take another look sometime in the next few days. For now, tests will fail for that.

### CODE REVIEW

**Comment by junpenglao** [comment]
> If I'm not mistaken, no other sampler computes the logp (e.g. Metropolis doesn't), so there would be overhead to require these samplers to track the logp. However, if this is a feature the dev team wants, I'd be happy to do that too - probably in a separate PR, though.

Oh you are right - in Metropolis we only computed the delta_logp. Hmmm, it does complicate things...

**Comment by eigenfoo** [comment]
This seems to pass tests now. The test it fails is due to SMC and appears unrelated. Ready for review! 

What I've done for CompoundStep is to remove `logp` items from all but the last `state` in `states`. In other words, only the last state can be the model logp. Sometimes, the last state may not have a logp at all (eg if the last sampler is Metropolis), but if the last state has a logp, then it is the model logp. Does that sound reasonable?

**Comment by junpenglao** [comment]
Need a new test to check if the logp is saved correctly at least.

Also, add a line to release-note.

I think this is a good first step. Maybe in another PR we can also add it to other samplers. I have an idea to do this also for Metropolis where we are computing the delta_logp instead of the logp. We can compute the model logp at sample 0, then from sample 1 onward we do `state.logp = logp_tm1 - delta_logp` which should recover the logp at time t while also cheap to compute.
Careful tests are of course needed to make sure it is correct especially for CompoundSteps.

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -3,6 +3,13 @@
 
 ## PyMC3 3.6
 
+### New features
+
+- Track the model log-likelihood as a sampler stat for NUTS and HMC samplers
+  (accessible as `trace.logp`).
+
+### Fixes
```
Comment (twiecki) [inline RELEASE-NOTES.md]: This is more maintenance.

**Comment by eigenfoo** [comment]
@junpenglao I'm actually not sure where the tests for sampler stats are. I would've thought `test_stats.py`, but that's actually to test the `pm.stats` module... Any pointers?

**Comment by junpenglao** [comment]
pymc3/tests/test_step.py

**Comment by eigenfoo** [comment]
Sorry for the lull, I was taking a break from my laptop!

I've added a test to assert the existence of various sampler stats for NUTS, and also assert that their shapes are correct. Let's see if this passes tests.

**Comment by eigenfoo** [comment]
Passes tests! @junpenglao ready for review

File: pymc3/step_methods/compound.py [hunk 2]
```diff
@@ -27,6 +27,10 @@ def step(self, point):
                     states.extend(state)
                 else:
                     point = method.step(point)
+            # Model logp can only be the logp of the _last_ state,
+            # if there is one. Remove all others.
+            for state in states[:-1]:
+                state.pop('logp', None)
```
Comment (ColCarroll) [inline pymc3/step_methods/compound.py]: i guess `states` here is an array of dictionaries? it is a `namedtuple` elsewhere, which doesn't have the `pop` method.

Comment (eigenfoo) [reply pymc3/step_methods/compound.py]: Oh hm that's tricky. The small-brain solution would be to check if `state` is a `namedtuple` or `dict` and remove `logp` appropriately... but that seems really inelegant. What do you think?

Also, is there any plan to make all `state`s either a dictionary or a tuple? It seems clunky to have them be different things...

**Comment by junpenglao** [comment]
Maybe we should rename it to `lp__`, see https://github.com/arviz-devs/arviz/pull/176#issuecomment-416112174

File: pymc3/tests/test_step.py [hunk 3]
```diff
@@ -456,3 +456,16 @@ def test_linalg(self, caplog):
...
+
+    def test_sampler_stats(self):
+        with Model():
+            x = Normal('x', mu=0, sd=1)
+            trace = sample(draws=10, tune=1, chains=1)
+
+        expected_stat_names = {'depth', 'diverging', 'energy', 'energy_error',
+            'logp', 'max_energy_error', 'mean_tree_accept', 'step_size',
+            'step_size_bar', 'tree_size', 'tune'}
+
```
Comment (junpenglao) [inline pymc3/tests/test_step.py]: Should also add test to check if the saved logp is the same compare to computing the logp post sample via passing the point from traces to logp_func

**Comment by eigenfoo** [comment]
Hm, my concern is how intuitive a name `lp__` is... I understand Stan has it as `lp__` but with Python there is the convention of having self-descriptive names, even if this means verbosity. I'd be fairly confused at both what `lp` stands for, and why there are two underscores after it (AFAIK nothing else in PyMC3 has two underscores after it...)

I don't really know much about Arviz though. Is "compatibility" with Stan/Arviz a high priority for the PyMC devs?

**Comment by twiecki** [comment]
I agree with @eigenfoo, `lp__` is not a great name. Compatibility to Arviz is important, but not to Stan. But since we also have control over arviz I would suggest standardizing on a reasonable name choice there (like `logp`) and here.

**Comment by junpenglao** [comment]
In that case, either `logp` or `model_logp` would be a good name choice.
FYI @eigenfoo, in PyMC3 the transformed parameters all have `__` which is generated internally. These RVs are meant to be "hidden" and we use the double underscore to identify these parameters and hide them when generating summary and figure.

**Comment by ColCarroll** [comment]
Yeah, agree on using a descriptive name - I think `arviz` is going to keep trying to decide on a "standard" name for sampler statistics, which may change over time (I'm still in favor of a descriptive name there as well, but it should be a friendly spot for `stan` users too). So no worries on what's going on there :)

File: pymc3/step_methods/compound.py (hunk 2 above)
Comment (eigenfoo) [reply pymc3/step_methods/compound.py]: Bump @ColCarroll

Comment (ColCarroll) [reply pymc3/step_methods/compound.py]: I'm not that familiar with `CompoundStep` - @junpenglao might have to speak to that.

 Let me take a look at what is going on, too: it makes me nervous that you need to _remove_ data in order to make the code work (also, a `namedtuple` has to have the entry still).

Comment (eigenfoo) [reply pymc3/step_methods/compound.py]: Sure thing! For now I'm going to go ahead with my suggestion: if it's a dictionary I'll pop it off, but if it's a namedtuple I'll set the `logp` to `np.nan`.

Comment (junpenglao) [reply pymc3/step_methods/compound.py]: Hmm, I think the best option is actually to record only the last logp (if there is any): `model_logp.append(states[-1]['logp'])`

File: pymc3/tests/test_step.py (hunk 3 above)
Comment (eigenfoo) [reply pymc3/tests/test_step.py]: I see, where exactly is `logp_func` and a list of the model's states? I can't find them in `model`...

**Comment by eigenfoo** [comment]
@junpenglao I think I like `model_logp` better - it's more explicit, and harder to confuse it with all the other log-probabilities that you could talk about 🙂

I'll need some help with the test though - see my comment above.

File: pymc3/tests/test_step.py (hunk 3 above)
Comment (junpenglao) [reply pymc3/tests/test_step.py]: What i mean is to compare the saved logp with:
```python
logp = model.logp
logp_ = np.array([logp(trace.point(i,chain=c)) for c in trace.chains for i in range(len(trace))]) 
```
as they should be the same. The same test could also be use when we implemented recording of logp for `CompoundStep`.

File: pymc3/step_methods/compound.py (hunk 2 above)
Comment (eigenfoo) [reply pymc3/step_methods/compound.py]: @junpenglao I don't think I understand... `model_logp` is a sampler stat and so must be carried around in `State`s (which are either dicts or tuples). It looks like you want it to be a list? Sorry for the confusion

File: pymc3/tests/test_step.py (hunk 3 above)
Comment (eigenfoo) [reply pymc3/tests/test_step.py]: Added!

File: pymc3/step_methods/compound.py (hunk 2 above)
Comment (junpenglao) [reply pymc3/step_methods/compound.py]: Yes but at the end the states are saved in the trace, with same length (nchain*nsample) as the posterior samples. So I would say if you save them separately it is ok as well - as long as you put them back into sampler stat before outputing it to user.

**Comment by springcoil** [comment]
Awesome work @eigenfoo

**Comment by aseyboldt** [comment]
Maybe better `sample_logp`? It isn't really a property of the model, and it is different for each sample. Or maybe `posterior_logp`? Otherwise I think this looks good.

**Comment by fonnesbeck** [comment]
It gets a little semantic, I suppose. Its the log-probability of the model at a particular point, so I don't think `model_logp` is necessarily bad. The sample doesn't have a logp without a model!

**Comment by aseyboldt** [comment]
How about `value_that_is_proportional_to_the_posterior_logp_given_model_and_dataset_at_the_point_of_that_sample`. We have tab completion after all. :-)
*Edit*: I forgot the parametrization. It depends on that as well.

**Comment by aseyboldt** [comment]
But I guess you have a point. They all probably work. (except that last one)

**Comment by ColCarroll** [comment]
This looks good to me (and something I will use right away) - thanks, @eigenfoo! Is there something specific you would still like review on?

**Comment by eigenfoo** [comment]
For the record, the problem that stalled me up was this:

The way `CompoundStep` works is essentially a glorified for-loop over a list of step methods. This has various implications for sampler stats, since each step method in this list may or may not support stats, and may support different stats, etc. etc. According to [these docs](https://docs.pymc.io/notebooks/sampler-stats.html#Multiple-samplers), we should be returning _every single stat_ generated by _every single sampler_. If there happen to be name collisions, we stick the numpy arrays together, along a new axis.

The "bad thing" about this PR is best explained by example:

```python
with pm.Model():
    x = pm.Normal('x')
    y = pm.Normal('y')
    step1 = pm.NUTS([mu1])
    step2 = pm.NUTS([mu2])
    trace = pm.sample(step=[step1, step2])
```

Here, `trace.get_sampler_stats('model_logp').shape` is `(1000, 2)`. This may be counter-intuitive, since there should only be _one_ model logp, so it _may_ make sense to insist that `model_logp` is always a 1-D numpy array.

Eventually, @ColCarroll and I decided that it's better to keep this quirk and document it, rather than having an inconsistent API. This means that if you wanted the overall logp for with a `CompoundStep`, you'd need `trace.get_sampler_stats('model_logp')[:, -1]`, and if you (for whatever reason) wanted one of the intermediate logp's, you'd need to index one of the columns of `trace.get_sampler_stats('model_logp')`.

I think we're going to go ahead and merge this once tests pass.

**Comment by twiecki** [comment]
Hm yeah that's unfortunate but I agree with your assessment.

**Comment by aseyboldt** [comment]
Thank @eigenfoo!

**Comment by ColCarroll** [comment]
Thanks for this!

**Pull request merged**